### ISSUE TITLE
Reset Password Token Expired

### ISSUE THREAD

**Issue by jestinepaul** [wekan/wekan#1062]
Followed the steps in production deployment. No mail server enabled.

Copied the link from docker logs, it gives the following "Token expired" error always.

![token-expired](https://user-images.githubusercontent.com/181524/27082529-c1f8ba20-5077-11e7-9cf7-08f7b106756e.png)

**Comment by connorads** [comment]
Hi @jestinepaul, I'm experiencing the same issue. Did you find reasonable a workaround?

**Comment by connorads** [comment]
I think I know what is causing the issue now.

For some reason, in the docker log there is  an additional `=` towards the end of the token in the password reset URL.
If you remove the `=` it should work.
![image](https://user-images.githubusercontent.com/10026538/29219480-17724776-7eb0-11e7-8f11-593dc206afca.png)
For me the additional `=` appears just before the start of a new line. Perhaps there is some "new line wrapping logic" somewhere which is appending an equals sign. Would be interesting to know if we get the same behavior when not using Docker.

There is an alternative way of getting the password reset token - from the database.
1. Request a password reset on the Wekan login screen
2. Login to your MongoDB and find the user that you just reset in the `users` collection
`db.getCollection('users').find({username: "insertyourusernamehere"})`
3. Look through the document to find the password reset token
Services > Password > Reset > Token
4. Append this token onto the password reset URL and it should work
i.e. http://yourwekaninstance/reset-password/pasteyourtokenhere

It would be nice if the message wasn't always "Token expired", even when you enter a completely invalid token.
Not sure what the security implications would be of not doing this or how easy this would be to implement, I'm assuming most of the heavy lifting is done in Meteor rather than Wekan.

**Comment by rautamiekka** [comment]
^ Is that screenshot from a Command Prompt window (on Window$) or from an external (outside Command Prompt) SSH Client ? Command Prompt, unlike NIX shells and external SSH Clients implementing such behaviour, creates a whole new line, which breaks single-line links and similar. The `5DNZ` on the following line makes me believe that's the case.

**Comment by connorads** [comment]
Hi @rautamiekka, the screenshot is taken from PuTTY.

You can see from the below screenshot that other messages from the same `docker logs` output are not being wrapped at the same width as the URL. This would lead me to believe that it isn't the terminal causing the issue.
![image](https://user-images.githubusercontent.com/10026538/29225039-ae7ec5d6-7ec4-11e7-99bd-4eb539a4f1da.png)

Also it's not just adding a new line, it's adding the `=` before the new line. Which is not normal line wrapping behavior for any application I know of ...

:mag: \*A few Google searches later\* :mag:

Looks a lot like this "equals adding line wrapping stuff" is a thing called [Quoted Printable Encoding](https://en.wikipedia.org/wiki/Quoted-printable).
> Quoted-Printable, or QP encoding, is an encoding using printable ASCII characters (alphanumeric and the equals sign =) to transmit 8-bit data over a 7-bit data path or, generally, over a medium which is not 8-bit clean.[1] It is defined as a MIME content transfer encoding for use in e-mail.

If this is the case, I wonder if there's a simple way to make it note use this encoding when the emails are output into the Wekan log.

**Comment by rautamiekka** [comment]
^ You're right, it's not wrapped, so that's 1 thing ruled out.

**Comment by connorads** [comment]
I'm pretty sure that the cause of this issue is because of the email encoding being used is `quoted-printable` as mentioned before.

## TL;DR
We should probably close this issue and forget about it!

### How could we go about fixing it?
It looks like it would be technically possible to change the `Content-Transfer-Encoding` when sending [emails with Meteor](https://docs.meteor.com/api/email.html) so that it's not `quoted-printable`. There doesn't appear to be a simple option with Meteor's mail package to change the encoding. However, the mail package is based on Mailcomposer and the [Mailcomposer documentation](https://nodemailer.com/extras/mailcomposer/#e-mail-message-fields) says that you can change it there. But this would obviously add a fair bit of complexity as it looks like you need to build your own `Mailcomposer` object and throw away what Meteor's `Email.send(options)` let's you set normally.

It's not actually clear if changing the encoding would fix the issue but it would appear by reading this [Stackoverflow Answer explaining the types of Content-Transfer-Encoding](https://stackoverflow.com/a/28531705) that changing it to `7bit` might work.

### Should we "fix" this issue?
Probably not, for many reasons.

Given that this issue is only experienced when the `MAIL_URL` environment variable has not been set, which is usually only when testing, I'd say that this issue could probably be closed without further action.
Also, this issue really has more to do with Meteor than it does with Wekan.

This issue is kind of user error really. These "trouble causing" password reset emails (or any other email) that is being output into the log are actually valid given that they are using the `quoted-printable` type of `Content-Transfer-Encoding`. But because it's being output into the log and not in an actual email, the encoding is visible to the user but not necessarily obvious to the untrained eye. The added `=` for the line break looks like it could be part of a valid password reset token, which is an unfortunate coincidence.

Hopefully anyone experiencing this problem in the future will find this GitHub issue and be able to use one of the workarounds in my post above.

### Additional findings
Here's an interesting one for you.
If you set your `ROOT_URL` to something short like `localhost`  it decides to use `7bit` encoding.
You won't get the added `=` issue because the URL doesn't exceed 78 characters and doesn't do a line wrap and add the `=`.
![image](https://user-images.githubusercontent.com/10026538/29293928-dcd2097a-8144-11e7-8498-372337d12125.png)

However if you choose a long `ROOT_URL`  you will get `quoted-printable` encoding.
You'll get the add `=`  issue because the URL exceeds 78 characters. You will get a line wrap and the additional `=`.
![image](https://user-images.githubusercontent.com/10026538/29294036-4a0516c2-8145-11e7-9626-4e20db9d173c.png)

The [Mailcomposer documentation](https://nodemailer.com/extras/mailcomposer/#e-mail-message-fields) seems to suggest it adjusts the encoding based on the email content type, perhaps this is why this happening.

### References
[Documentation of Meteor's email API](https://docs.meteor.com/api/email.html)
[Mailcomposer Documentation](https://nodemailer.com/extras/mailcomposer/#e-mail-message-fields)
[Stackoverflow Answer explaining the types of Content-Transfer-Encoding](https://stackoverflow.com/a/28531705)

PS. Why did I write this essay 😕

**Comment by rautamiekka** [comment]
^
> PS. Why did I write this essay 😕

That's not encouraging as far as development is concerned.

**Comment by connorads** [comment]
Hey @xet7 
What do you think? Should we close this issue?

**Comment by xet7** [comment]
@connorads 

Thanks a lot, I welcome all essays that result in closing issue, and all info about issues that can be closed! 👍
